File: packages/flags-python/src/signakit_flags/hasher.py
"""Deterministic hashing for consistent user bucketing.

Implements MurmurHash3 32-bit, byte-for-byte compatible with the JS SDK
(``packages/flags-node/src/hasher.ts``) and PHP SDK (``Hasher.php``) for
ASCII-safe inputs (which is the common case for ``salt:userId``).
"""

from __future__ import annotations

from .constants import BUCKET_SPACE

_C1 = 0xCC9E2D51
_C2 = 0x1B873593
_R1 = 15
_R2 = 13
_M = 5
_N = 0xE6546B64
_MASK32 = 0xFFFFFFFF


def _rotl32(x: int, r: int) -> int:
    """Rotate ``x`` left by ``r`` bits within a 32-bit unsigned word."""
    return ((x << r) | (x >> (32 - r))) & _MASK32


def _fmix32(h: int) -> int:
    """MurmurHash3 32-bit finalisation mix."""
    h ^= h >> 16
    h = (h * 0x85EBCA6B) & _MASK32
    h ^= h >> 13
    h = (h * 0xC2B2AE35) & _MASK32
    h ^= h >> 16
    return h & _MASK32


def murmur3_32(key: str, seed: int = 0) -> int:
    """Compute MurmurHash3 32-bit of ``key`` (UTF-8) with the given seed.

    Args:
        key: Input string (encoded as UTF-8 bytes for hashing).
        seed: 32-bit seed value (default 0).

    Returns:
        Unsigned 32-bit hash value (``0`` – ``4_294_967_295``).
    """
    data = key.encode("utf-8")
    length = len(data)
    nblocks = length // 4

    h1 = seed & _MASK32

    # ---- body: process 4-byte little-endian blocks ----
    for i in range(nblocks):
        offset = i * 4
        k1 = (
            data[offset]
            | (data[offset + 1] << 8)
            | (data[offset + 2] << 16)
            | (data[offset + 3] << 24)
        )

        k1 = (k1 * _C1) & _MASK32
        k1 = _rotl32(k1, _R1)
        k1 = (k1 * _C2) & _MASK32

        h1 ^= k1
        h1 = _rotl32(h1, _R2)
        h1 = ((h1 * _M) & _MASK32) + _N
        h1 &= _MASK32

    # ---- tail: 1–3 remaining bytes ----
    tail_index = nblocks * 4
    remain = length - tail_index
    k1 = 0

    if remain == 3:
        k1 ^= data[tail_index + 2] << 16
        k1 ^= data[tail_index + 1] << 8
        k1 ^= data[tail_index]
    elif remain == 2:
        k1 ^= data[tail_index + 1] << 8
        k1 ^= data[tail_index]
    elif remain == 1:
        k1 ^= data[tail_index]

    if remain > 0:
        k1 = (k1 * _C1) & _MASK32
        k1 = _rotl32(k1, _R1)
        k1 = (k1 * _C2) & _MASK32
        h1 ^= k1

    # ---- finalisation ----
    h1 ^= length
    h1 = _fmix32(h1)
    return h1


def hash_to_bucket(salt: str, user_id: str) -> int:
    """Hash ``salt:user_id`` into a bucket in ``[0, BUCKET_SPACE)``."""
    return murmur3_32(f"{salt}:{user_id}") % BUCKET_SPACE


def hash_for_traffic(salt: str, user_id: str) -> int:
    """Bucket for traffic allocation — uses the ``:traffic`` namespace."""
    return hash_to_bucket(f"{salt}:traffic", user_id)


def hash_for_variation(salt: str, user_id: str) -> int:
    """Bucket for variation allocation — uses the ``:variation`` namespace."""
    return hash_to_bucket(f"{salt}:variation", user_id)


def hash_for_default(salt: str, user_id: str) -> int:
    """Bucket for default allocation — uses the ``:default`` namespace."""
    return hash_to_bucket(f"{salt}:default", user_id)

File: packages/flags-php/src/Hasher.php
<?php

declare(strict_types=1);

namespace SignaKit\FlagsPhp;

/**
 * MurmurHash3 32-bit implementation matching the JS SDK exactly.
 *
 * PHP integers are 64-bit on 64-bit platforms, so every multiplication is
 * masked with `& 0xFFFFFFFF` to simulate 32-bit unsigned overflow.
 * The final modulo uses fmod() to avoid integer overflow on very large hashes.
 */
final class Hasher
{
    private const BUCKETS = 10000;

    /**
     * Compute MurmurHash3 32-bit for a UTF-8 string with seed 0.
     *
     * @param string $key Input string
     * @return int Unsigned 32-bit hash value (0 – 4294967295)
     */
    public static function murmur3_32(string $key): int
    {
        $data   = $key;
        $length = strlen($data);
        $nblocks = (int) ($length / 4);

        $h1 = 0;

        $c1 = 0xcc9e2d51;
        $c2 = 0x1b873593;

        // ---- body: process 4-byte blocks ----
        for ($i = 0; $i < $nblocks; $i++) {
            $k1 = self::getInt32($data, $i * 4);

            $k1 = ($k1 * $c1) & 0xFFFFFFFF;
            $k1 = self::rotl32($k1, 15);
            $k1 = ($k1 * $c2) & 0xFFFFFFFF;

            $h1 ^= $k1;
            $h1 = self::rotl32($h1, 13);
            $h1 = (($h1 * 5) & 0xFFFFFFFF) + 0xe6546b64;
            $h1 &= 0xFFFFFFFF;
        }

        // ---- tail: remaining 1–3 bytes ----
        $tail   = $nblocks * 4;
        $remain = $length - $tail;
        $k1     = 0;

        switch ($remain) {
            case 3:
                $k1 ^= ord($data[$tail + 2]) << 16;
                // fall through
            case 2:
                $k1 ^= ord($data[$tail + 1]) << 8;
                // fall through
            case 1:
                $k1 ^= ord($data[$tail]);
                $k1  = ($k1 * $c1) & 0xFFFFFFFF;
                $k1  = self::rotl32($k1, 15);
                $k1  = ($k1 * $c2) & 0xFFFFFFFF;
                $h1 ^= $k1;
        }

        // ---- finalisation ----
        $h1 ^= $length;
        $h1  = self::fmix32($h1);

        return $h1;
    }

    /** Traffic-allocation hash: returns a bucket 0–9999 */
    public static function hashForTraffic(string $salt, string $userId): int
    {
        return (int) fmod(self::murmur3_32("{$salt}:traffic:{$userId}"), self::BUCKETS);
    }

    /** Variation-allocation hash: returns a bucket 0–9999 */
    public static function hashForVariation(string $salt, string $userId): int
    {
        return (int) fmod(self::murmur3_32("{$salt}:variation:{$userId}"), self::BUCKETS);
    }

    /** Default-allocation hash: returns a bucket 0–9999 */
    public static function hashForDefault(string $salt, string $userId): int
    {
        return (int) fmod(self::murmur3_32("{$salt}:default:{$userId}"), self::BUCKETS);
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    /**
     * Read a little-endian 32-bit integer from the byte string.
     */
    private static function getInt32(string $data, int $offset): int
    {
        return (
            (ord($data[$offset])       )        |
            (ord($data[$offset + 1])    <<  8)   |
            (ord($data[$offset + 2])    << 16)   |
            (ord($data[$offset + 3])    << 24)
        ) & 0xFFFFFFFF;
    }

    /**
     * Rotate left 32-bit.
     */
    private static function rotl32(int $x, int $r): int
    {
        return (($x << $r) | ($x >> (32 - $r))) & 0xFFFFFFFF;
    }

    /**
     * MurmurHash3 32-bit finalisation mix (fmix32).
     */
    private static function fmix32(int $h): int
    {
        $h ^= ($h >> 16);
        $h  = ($h * 0x85ebca6b) & 0xFFFFFFFF;
        $h ^= ($h >> 13);
        $h  = ($h * 0xc2b2ae35) & 0xFFFFFFFF;
        $h ^= ($h >> 16);

        return $h & 0xFFFFFFFF;
    }
}
